pkg/config/viper.go

```diff
@@ -16,243 +16,267 @@ import (
 func (c *safeConfig) GetInt(key string) int {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetInt(key)
 }
 
-// GetInt64 is wrapped for concurrent access
+// GetInt64 wraps Viper for concurrent access
 func (c *safeConfig) GetInt64(key string) int64 {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetInt64(key)
 }
 
-// GetFloat64 is wrapped for concurrent access
+// GetFloat64 wraps Viper for concurrent access
 func (c *safeConfig) GetFloat64(key string) float64 {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetFloat64(key)
 }
 
-// GetTime is wrapped for concurrent access
+// GetTime wraps Viper for concurrent access
 func (c *safeConfig) GetTime(key string) time.Time {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetTime(key)
 }
 
-// GetDuration is wrapped for concurrent access
+// GetDuration wraps Viper for concurrent access
 func (c *safeConfig) GetDuration(key string) time.Duration {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetDuration(key)
 }
 
-// GetStringSlice is wrapped for concurrent access
+// GetStringSlice wraps Viper for concurrent access
 func (c *safeConfig) GetStringSlice(key string) []string {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringSlice(key)
 }
 
-// GetStringMap is wrapped for concurrent access
+// GetStringMap wraps Viper for concurrent access
 func (c *safeConfig) GetStringMap(key string) map[string]interface{} {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringMap(key)
 }
 
-// GetStringMapString is wrapped for concurrent access
+// GetStringMapString wraps Viper for concurrent access
 func (c *safeConfig) GetStringMapString(key string) map[string]string {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringMapString(key)
 }
 
-// GetStringMapStringSlice is wrapped for concurrent access
+// GetStringMapStringSlice wraps Viper for concurrent access
 func (c *safeConfig) GetStringMapStringSlice(key string) map[string][]string {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringMapStringSlice(key)
 }
 
-// GetSizeInBytes is wrapped for concurrent access
+// GetSizeInBytes wraps Viper for concurrent access
 func (c *safeConfig) GetSizeInBytes(key string) uint {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetSizeInBytes(key)
 }
 
-// SetEnvPrefix is wrapped for concurrent access
+// SetEnvPrefix wraps Viper for concurrent access, and keeps the envPrefix for
+// future reference
 func (c *safeConfig) SetEnvPrefix(in string) {
 	c.Lock()
 	defer c.Unlock()
 	c.Viper.SetEnvPrefix(in)
+	c.envPrefix = in
 }
 
-// BindEnv is wrapped for concurrent access
+// BindEnv wraps Viper for concurrent access, and adds tracking of the configurable env vars
 func (c *safeConfig) BindEnv(input ...string) error {
 	c.RLock()
 	defer c.RUnlock()
+	if len(input) == 1 {
+		key := input[0]
+		if !strings.Contains(key, "_key") {
+			envVarName := strings.Join([]string{c.envPrefix, strings.ToUpper(key)}, "_")
+			c.configEnvVars = append(c.configEnvVars, envVarName)
```

Comment:
Shouldn't the `c.RLock()` be changed to a `c.Lock()` ?

Reply:
the lock was originally only meant to make calls to viper thread-safe, but you're right in this case we should fully `Lock`